src/components/modules/caret.ts

```diff
@@ -548,6 +558,117 @@ export default class Caret extends Module {
     selection.addRange(newRange);
   }
 
+  /**
+   * Detect an next line position from the caret position
+   */
+  private detectNextLinePosition(): Position | false {
+    const { BlockManager, Tools } = this.Editor;
+
+    const nextInput = BlockManager.currentBlock.nextInput;
+    let nextBlock = BlockManager.nextContentfulBlock;
+
+    if (!nextBlock && !nextInput) {
+      /**
+       * If there is a last block and it is default, do nothing
+       * This code allows to exit from the last non-initial tool:
+       * https://github.com/codex-team/editor.js/issues/1103
+       */
+      if (BlockManager.currentBlock.tool.isDefault) {
+        return false;
+      }
+
+      /**
+       * If there is no nextBlock, but currentBlock is not default,
+       * insert new default block at the end and navigate to it
+       */
+      nextBlock = BlockManager.insertAtEnd();
+    }
+
+    return this.detectPosition('next', nextBlock, nextInput);
+  }
+
+  /**
+   * Detect a previous line position from the caret position
+   */
+  private detectPreviousLinePosition(): Position | false {
+    const { BlockManager } = this.Editor;
+
+    const previousInput = BlockManager.currentBlock.previousInput;
+    const previousBlock = BlockManager.previousContentfulBlock;
+
+    if (!previousBlock && !previousInput) {
+      return false;
+    }
+
+    return this.detectPosition('previous', previousBlock, previousInput);
+  }
+
+  /**
+   * Detect an next or previous line position from the caret position
+   *
+   * @param {'next' | 'previous'} direction - the direction of detection
+   * @param {Block} block - next or previous Block
+   * @param {HTMLElement | undefined} input - next or previous Tool's input
+   */
+  private detectPosition(direction: 'next' | 'previous', block: Block, input?: HTMLElement): Position {
+    const caretBoundingClientRect = Selection.get().getRangeAt(0)
+      .getBoundingClientRect();
+
+    const range = new Range();
+    const root = input ?? block.firstInput;
+
+    let offset = direction === 'next' ? 0 : root.textContent.length - 1;
+    let previousBoundingClientRect: DOMRect | undefined;
+
+    let position = {
+      block,
+      input,
+      offset: direction === 'next' ? root.textContent.length : 0,
+    };
+
+    /**
+     * Detect the nearest offset by horizontal position
+     */
+    this.walkTextNodeChars(root, direction, (textNode, index) => {
+      range.setStart(textNode, index);
+      range.setEnd(textNode, index + 1);
+
+      const currentBoundingClientRect = range.getBoundingClientRect();
+
+      if (previousBoundingClientRect) {
+        const caretX = caretBoundingClientRect.x + (direction === 'previous' ? caretBoundingClientRect.width : 0);
+        const currentX = currentBoundingClientRect.x + (direction === 'previous' ? currentBoundingClientRect.width : 0);
+        const previousX = previousBoundingClientRect.x + (direction === 'previous' ? previousBoundingClientRect.width : 0);
```

Comment:
`(direction === 'previous' ? previousBoundingClientRect.width : 0)` could be extracted to a method

```
const getXOffset = (rectWidth: number) => direction === 'previous' ? rectWidth : 0;
```